experimental/execution/ptx-executor/src/scheduler.rs

```diff
@@ -0,0 +1,275 @@
+        value: Option<StateValue>,
+    },
+    AddTransaction {
+        txn_idx: TxnIdx,
+        transaction: Transaction,
+        dependencies: HashSet<(StateKey, TxnIdx)>,
+    },
+    FinishBlock,
+    Exit,
+}
+
+enum StateValueState {
+    Pending { subscribers: Vec<TxnIdx> },
+    Ready { value: Option<StateValue> },
+}
+
+struct PendingTransaction {
+    transaction: Transaction,
+    pending_dependencies: HashSet<VersionedKey>,
+    met_dependencies: HashMap<StateKey, Option<StateValue>>,
+}
+
+impl PendingTransaction {
+    fn new(transaction: Transaction) -> Self {
+        Self {
+            transaction,
+            pending_dependencies: HashSet::new(),
+            met_dependencies: HashMap::new(),
+        }
+    }
+}
+
+struct Worker {
+    transactions: Vec<Option<PendingTransaction>>,
+    state_values: HashMap<VersionedKey, StateValueState>,
+    num_pending_txns: usize,
+    block_finished: bool,
+    runner: PtxRunnerClient,
+    scheduler: PtxSchedulerClient,
+}
+
+impl Worker {
+    fn new(runner: PtxRunnerClient, scheduler: PtxSchedulerClient) -> Self {
+        Self {
+            transactions: Vec::with_capacity(EXPECTANT_BLOCK_SIZE),
+            state_values: HashMap::with_capacity(EXPECTANT_BLOCK_KEYS),
+            num_pending_txns: 0,
+            block_finished: false,
+            runner,
+            scheduler,
+        }
+    }
+
+    pub fn inform_state_value(&mut self, key: VersionedKey, value: Option<StateValue>) {
+        let _timer = TIMER.timer_with(&["scheduler_inform_state_value"]);
+        match self.state_values.entry(key.clone()) {
+            Entry::Occupied(mut existing) => {
+                let old_state = existing.insert(StateValueState::Ready {
+                    value: value.clone(),
+                });
+                match old_state {
+                    StateValueState::Pending { subscribers } => {
+                        for subscriber in subscribers {
+                            // TODO(ptx): reduce clone / memcpy
+                            self.inform_state_value_to_txn(subscriber, key.clone(), value.clone());
+                        }
+                    },
+                    StateValueState::Ready { .. } => {
+                        unreachable!("StateValue pushed twice.")
+                    },
+                }
+            },
+            Entry::Vacant(vacant) => {
+                vacant.insert(StateValueState::Ready { value });
+            },
+        }
+    }
+
+    fn inform_state_value_to_txn(
+        &mut self,
+        txn_idx: TxnIdx,
+        versioned_key: VersionedKey,
+        value: Option<StateValue>,
+    ) {
+        let pending_txn = self.borrow_pending_txn(txn_idx);
+        let found_dependency = pending_txn.pending_dependencies.remove(&versioned_key);
+        assert!(found_dependency, "Pending dependency not found.");
+        let (key, _txn_idx) = versioned_key;
+        pending_txn.met_dependencies.insert(key, value);
```

Comment:
are we assuming the pending_dependencies won't have two keys with the same StateKey but different version?

Do we really need two hashmaps here instead of one hashmap and a counter?

Reply:
That'll need to be `HashMap<StateKey, Option<Option<StateVaule>>>`, but yeah, doable, not sure if it's worth it. (current version looks straightforward to me)